recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl

```diff
@@ -761,113 +955,97 @@ def main(argv):
-            )
-            if not preserved_env_input == "n":
-                updater.update_uboot_env(envlist)
+            updater.backup()
 
-        reboot_input = the_input(
-            args.quiet, 'n',
-            "\n" + BOLD + "Warning: Completed. Please reboot the device."
-            + ENDC + " Reboot now (y/N)? "
-        )
-        if not reboot_input == "y":
-            return ErrorCode.SUCCESS.value
-        os.system('reboot')
+        index = 0
+        while True:
+            try:
+                updater.update()
+                break
+            except UpgradeError as e:
+                if index > 2:
+                    raise UpgradeError(e.err, e.code)
+                index += 1
+                print("{}, try again!".format(e.err))
     except UpgradeError as e:
-        print(e)
-        intput_reminder = '''
-        ====================================
-        Upgrade Failed! Please do not reboot!!
-        The Device may become brick.
-        Please try to upgrade firmware again!
-        ====================================
+        print(e.err)
+        input_reminder = '''
+   ====================================
+   Upgrade Failed! Please do not reboot!!
+   The Device may become brick.
+   Please try to rollback!
+   ====================================
 Hit the Enter Key to Exit:
-        '''
-        # Rollback the firmware when the upgrade fails
-        try:
-            print(BOLD + 'Warning: Upgrade failed, rollback started' +
-                  ENDC + '\n')
-            updater = RollbackFirmware(args.firmware, args.backup_dir)
-            if ErrorCode.ROLLBACK_FAILED.value == updater.rollback_firmware():
-                return ErrorCode.ROLLBACK_FAILED.value
-        except (UpgradeError, UpgradeRollbackError) as e:
-            print(e)
-
-        the_input(args.quiet, " ", intput_reminder)
-        return ErrorCode.FLASHING_FAILED.value
-    except UpgradeRollbackError as e:
-        print(e)
-        intput_reminder = '''
-        ====================================
-        Rollback Failed! Please do not reboot!!
-        The Device may become brick.
-        Please try to rollback firmware again!
-        ====================================
-Hit the Enter Key to Exit:
-        '''
+'''
+        if e.code == ErrorCode.FLASHING_FAILED.value:
+            interactor.interact(" ", input_reminder)
+        if args.rollback:
+            e.code = ErrorCode.ROLLBACK_FAILED.value
+
+        return e.code
+
+    reboot_input = interactor.interact(
+        'n',
+        "\nWarning: Completed. Please reboot the device. Reboot now (y/N)?"
+    )
+    if not reboot_input == "y":
+        return ErrorCode.SUCCESS.value
+    os.system('reboot')
 
-        the_input(args.quiet, " ", intput_reminder)
-        return ErrorCode.ROLLBACK_FAILED.value
 
 if __name__ == '__main__':
-    code = main(sys.argv)
-    sys.exit(code)
+    CODE = main(sys.argv)
+    sys.exit(CODE)
```

Comment:
So, does all this change the logic of the update script in any way or really just the internal structure? Commit message is not fully clear on this, and I'm too lazy :wink: to read all the above.

Reply:
This only changes the internal structure, no notable functional/logic changes, maybe some minor changes, but that's trivial.

Reply:
I will highlight this into commit message.